### ISSUE TITLE
Crash on Windows 7 x86 in waveOutOpen

### ISSUE THREAD

**Issue by divVerent** [hajimehoshi/oto#158]
There seems a C-side NULL dereference in this line:

https://github.com/hajimehoshi/oto/blob/main/winmm_windows.go#L136

Details and backtrace are here: https://github.com/divVerent/aaaaxy/issues/17

Could reproduce on a freshly installed Windows 7 VM. The exact same binary is fine in WINE and in Windows 10, though.

Note that this error implies that the DLL function at least does exist - otherwise LazyDLL would have panicked. Something instead must be wrong with its arguments.

Go itself still officially supports Windows 7 (even though Microsoft no longer does), so it would at least be interesting to figure out what this is.

**Comment by divVerent** [comment]
Caught it in OllyDbg now.

![image](https://user-images.githubusercontent.com/251568/153331318-9c29280e-9cb2-4f96-b20c-26434c345640.png)

You see the six arguments to waveOutOpen:

```
  LPHWAVEOUT      phwo,  // 0x1408BA78, seems OK
  UINT            uDeviceID,  // 0x35729858, seems OK
  LPCWAVEFORMATEX pwfx,  // 0x13C89650, seems OK
  DWORD_PTR       dwCallback,  // NULL?!?
  DWORD_PTR       dwInstance,  // NULL, is unused so OK
  DWORD           fdwOpen  // 0x00030000 == CALLBACK_FUNCTION
```

So for some reason we're passing a NULL callback.

**Comment by divVerent** [comment]
Odd, by reading oto code this is outright impossible: https://github.com/hajimehoshi/oto/blob/main/winmm_windows.go#L137

**Comment by divVerent** [comment]
Sorry, it's more complicated than that; upon entry of waveOutOpen the callback pointer is still there:

![image](https://user-images.githubusercontent.com/251568/153334555-196f8093-04f6-4108-a150-18fbe23be3af.png)

This value becomes zero, but I can't put a memory breakpoint on the stack across syscalls, so... wat.

**Comment by divVerent** [comment]
Wonder if it does a tail call if the device is -1 (WAVE_MAPPER)...? And somehow screws up the callback then?

**Comment by divVerent** [comment]
A breakpoint on waveOutOpen reveals a LOT of other calls to it, apparently trying out all device IDs. These all have flags 3 and no callback, and lie deeper in the stack (called by msacm32). After that sequence, the crash happens without any other calls to waveOutOpen.

**Comment by hajimehoshi** [comment]
> https://github.com/hajimehoshi/oto/blob/main/winmm_windows.go#L136

I realized that checking the error code (by GetLastError) might not make sense if the function succeeds. I'll update the implementation soon later.

**Comment by hajimehoshi** [comment]
Could you try `01e4dbe35a33e65bae78ad24c0974c16ca50cbf1`?

**Comment by divVerent** [comment]
Crash remains unchanged; the crash is really within oto's call to waveOutOpen, not after.

The other (not so related) issue I have noticed - no sound device = game won't start - also remains unchanged:

![image](https://user-images.githubusercontent.com/251568/153405364-de09a608-8dff-4701-a16b-0e7b0ab2af3f.png)

I wonder though if this issue here is even in oto or if something below oto is simply not compatible with Windows 7 anymore. Not sure how to best track that down though.

I also did observe that our callback MIGHT be doing more than allowed - after all, Go is a GC'd language, so who knows what might happen - and thus wonder if CALLBACK_EVENT would be a better option than CALLBACK_FUNCTION. For reference, inside a CALLBACK_FUNCTION, this restriction applies:

> Applications should not call any system-defined functions from inside a callback function, except for [EnterCriticalSection](https://msdn.microsoft.com/en-us/library/ms682608(v=vs.85)), [LeaveCriticalSection](https://msdn.microsoft.com/en-us/library/ms684169(v=vs.85)), [midiOutLongMsg](https://docs.microsoft.com/en-us/previous-versions/dd798474(v=vs.85)), [midiOutShortMsg](https://docs.microsoft.com/en-us/previous-versions/dd798481(v=vs.85)), [OutputDebugString](https://msdn.microsoft.com/en-us/library/aa363362(v=vs.85)), [PostMessage](https://msdn.microsoft.com/en-us/library/ms644944(v=vs.85)), [PostThreadMessage](https://msdn.microsoft.com/en-us/library/ms644946(v=vs.85)), SetEvent, [timeGetSystemTime](https://msdn.microsoft.com/en-us/library/dd757628(v=vs.85)), [timeGetTime](https://msdn.microsoft.com/en-us/library/dd757629(v=vs.85)), [timeKillEvent](https://docs.microsoft.com/en-us/previous-versions/dd757630(v=vs.85)), and [timeSetEvent](https://docs.microsoft.com/en-us/previous-versions/dd757634(v=vs.85)). Calling other wave functions will cause deadlock.

However, I can't tell IF our callback contains anything else; I am pretty sure they have at least the theoretical ability to panic (e.g. due to being out of stack space).

Can't now, but gotta find a way to check if the callback is run and maybe calls something forbidden while running.

**Comment by hajimehoshi** [comment]
Thank you for investigating. I'm fine to switch the callback to an event if it is really necessary. However, it seems a little awkward to use a native API for a condition variable.

**Comment by divVerent** [comment]
Found it.

After the callback has been called, note how the return address is still 9 words away:

![image](https://user-images.githubusercontent.com/251568/153519857-2ca8dca8-2fc0-4827-8b3a-d78f854dfc93.png)

However, there are only four `POP` before the `RET`. So it appears like the function in Windows 7 requires the callback to eat its arguments from the stack (via a `RET` instruction that after returning adds something to the stack pointer). Our callback however left the stack pointer alone, and expects the caller to clean up the stack.

In other words, this is a calling convention mismatch. Oto has a `__cdecl` callback, but we need a `__stdcall` one.

In MSDN docs, the function is declared as `CALLBACK`: https://docs.microsoft.com/en-us/previous-versions/dd743869(v=vs.85)

What does `CALLBACK` expand to? https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types says it is `__stdcall`. Ebiten, however, provided a `__cdecl` function.

Now the final mystery - why does the 32-bit binary then work at all on some machines? Looking at the DLL that ships with WINE solves the riddle:

```
6a4042f0 <_DriverCallback@28>:
[...]
6a40448d:       89 44 24 04             mov    DWORD PTR [esp+0x4],eax
6a404491:       ff d6                   call   esi  <-- THIS is our callback
6a404493:       0f b6 05 58 50 42 6a    movzx  eax,BYTE PTR ds:0x6a425058
6a40449a:       83 ec 14                sub    esp,0x14
6a40449d:       a8 08                   test   al,0x8
6a40449f:       0f 84 15 ff ff ff       je     6a4043ba <_DriverCallback@28+0xca>  <-- branch taken
[...]
6a4043ba:       8d 65 f0                lea    esp,[ebp-0x10]
6a4043bd:       89 d8                   mov    eax,ebx
6a4043bf:       59                      pop    ecx
6a4043c0:       5b                      pop    ebx
6a4043c1:       5e                      pop    esi
6a4043c2:       5f                      pop    edi
6a4043c3:       5d                      pop    ebp
6a4043c4:       8d 61 fc                lea    esp,[ecx-0x4]
6a4043c7:       c2 1c 00                ret    0x1c
```

Observe this `LEA` instruction that restores the stack pointer by subtracting 0x10 from `EBP`. In other words, the fact that the callback messed up the stack pointer is compensated here, kinda accidentally (the C code does not intend to do this, it's just what the compiler produced). I bet if I compiled the DLL with `-fomit-frame-pointer`, it'd break just like on Windows 7.

Sending patch to oto.
